plugins/mattermost/client/utils/mattermost.ts

```diff
@@ -0,0 +1,77 @@
+import { userDetailsUrl as userDataUrl, userTeamsUrl } from "../../shared/url";
+import { UserData, TeamData } from "./types";
+import { Dictionary } from "./useDictionary";
+
+export class MattermostClient {
+  #serverUrl: string;
+  #apiKey: string;
+  #dictionary: Dictionary;
+
+  constructor({
+    serverUrl,
+    apiKey,
+    dictionary,
+  }: {
+    serverUrl: string;
+    apiKey: string;
+    dictionary: Dictionary;
+  }) {
+    this.#serverUrl = serverUrl;
+    this.#apiKey = apiKey;
+    this.#dictionary = dictionary;
+  }
+
+  getUserData = async (): Promise<UserData> => {
+    const url = userDataUrl({ serverUrl: this.#serverUrl });
+    return this.#invokeMattermost<UserData>({
+      url,
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      parseResponse: (data: any) => ({
+        id: data.id,
+        name: data.username,
+        email: data.email,
+      }),
+    });
+  };
+
+  getUserTeams = async (): Promise<TeamData[]> => {
+    const url = userTeamsUrl({ serverUrl: this.#serverUrl });
+    return this.#invokeMattermost<TeamData[]>({
+      url,
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      parseResponse: (data: any) =>
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        data.map((team: any) => ({ id: team.id, name: team.display_name })),
+    });
+  };
+
+  #invokeMattermost = async <T>({
+    url,
+    parseResponse,
+  }: {
+    url: string;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    parseResponse: (json: any) => T;
+  }) => {
+    let res: Response;
+
+    try {
+      res = await fetch(url, {
```

Comment:
Alternatively we could create an endpoint for this, wouldn't that avoid the need to enable CORS on the Mattermost server?

Reply:
It'd introduce an additional network hop to our backend, but I think it makes it cleaner to have all mattermost interactions from our backend; will make this change.